File: app/src/main/java/me/zhanghai/android/files/ui/ClickableMovementMethod.java
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.ui;

import android.text.Layout;
import android.text.Selection;
import android.text.Spannable;
import android.text.method.BaseMovementMethod;
import android.text.method.LinkMovementMethod;
import android.text.style.ClickableSpan;
import android.view.MotionEvent;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * A movement method that traverses links in the text buffer and fires clicks. Unlike
 * {@link LinkMovementMethod}, this will not consume touch events outside {@link ClickableSpan}s.
 */
public class ClickableMovementMethod extends BaseMovementMethod {

    @Nullable
    private static ClickableMovementMethod sInstance;

    @NonNull
    public static ClickableMovementMethod getInstance() {
        if (sInstance == null) {
            sInstance = new ClickableMovementMethod();
        }
        return sInstance;
    }

    @Override
    public boolean canSelectArbitrarily() {
        return false;
    }

    @Override
    public boolean onTouchEvent(@NonNull TextView widget, @Nullable Spannable buffer,
                                @NonNull MotionEvent event) {

        int action = event.getActionMasked();
        if (action == MotionEvent.ACTION_UP || action == MotionEvent.ACTION_DOWN) {

            int x = (int) event.getX();
            int y = (int) event.getY();
            x -= widget.getTotalPaddingLeft();
            y -= widget.getTotalPaddingTop();
            x += widget.getScrollX();
            y += widget.getScrollY();

            Layout layout = widget.getLayout();
            int line = layout.getLineForVertical(y);
            int off = layout.getOffsetForHorizontal(line, x);

            ClickableSpan[] link = buffer.getSpans(off, off, ClickableSpan.class);
            if (link.length > 0) {
                if (action == MotionEvent.ACTION_UP) {
                    link[0].onClick(widget);
                } else {
                    Selection.setSelection(buffer, buffer.getSpanStart(link[0]),
                            buffer.getSpanEnd(link[0]));
                }
                return true;
            } else {
                Selection.removeSelection(buffer);
            }
        }

        return false;
    }

    @Override
    public void initialize(@NonNull TextView widget, @NonNull Spannable text) {
        Selection.removeSelection(text);
    }
}

File: app/src/main/java/me/zhanghai/android/files/util/ClickableMovementMethod.kt
/*
 * Copyright (c) 2020 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.util

import android.text.Selection
import android.text.Spannable
import android.text.method.BaseMovementMethod
import android.text.style.ClickableSpan
import android.view.MotionEvent
import android.widget.TextView

/**
 * A movement method that traverses links in the text buffer and fires clicks. Unlike
 * [android.text.method.LinkMovementMethod], this will not consume touch events outside
 * [ClickableSpan]s.
 */
object ClickableMovementMethod : BaseMovementMethod() {
    override fun initialize(view: TextView, text: Spannable) {
        Selection.removeSelection(text)
    }

    override fun onTouchEvent(view: TextView, text: Spannable, event: MotionEvent): Boolean {
        when (val action = event.actionMasked) {
            MotionEvent.ACTION_DOWN, MotionEvent.ACTION_UP -> {
                val x = event.x.toInt() - view.totalPaddingLeft + view.scrollX
                val y = event.y.toInt() - view.totalPaddingTop + view.scrollY
                val layout = view.layout
                val span = if (y < 0 || y > layout.height) {
                    null
                } else {
                    val line = layout.getLineForVertical(y)
                    if (x < layout.getLineLeft(line) || x > layout.getLineRight(line)) {
                        null
                    } else {
                        val off = layout.getOffsetForHorizontal(line, x.toFloat())
                        text.getSpans(off, off, ClickableSpan::class.java).firstOrNull()
                    }
                }
                if (span != null) {
                    if (action == MotionEvent.ACTION_DOWN) {
                        Selection.setSelection(text, text.getSpanStart(span), text.getSpanEnd(span))
                    } else {
                        span.onClick(view)
                    }
                    return true
                } else {
                    Selection.removeSelection(text)
                }
            }
        }
        return false
    }
}
